x/wasm-storage/types/filters.go

```diff
@@ -0,0 +1,299 @@
+// 0             1           9             10                 18 18+json_path_length
+// | filter_type | max_sigma | number_type | json_path_length | json_path |
+func NewFilterStdDev(input []byte) (FilterStdDev, error) {
+	var filter FilterStdDev
+	if len(input) < 18 {
+		return filter, ErrInvalidFilterInputLen.Wrapf("%d < %d", len(input), 18)
+	}
+
+	var maxSigma uint64
+	err := binary.Read(bytes.NewReader(input[1:9]), binary.BigEndian, &maxSigma)
+	if err != nil {
+		return filter, err
+	}
+	filter.maxSigma = maxSigma
+
+	filter.numberType = input[9]
+
+	var pathLen uint64
+	err = binary.Read(bytes.NewReader(input[10:18]), binary.BigEndian, &pathLen)
+	if err != nil {
+		return filter, err
+	}
+
+	path := input[18:]
+	if len(path) != int(pathLen) {
+		return filter, ErrInvalidPathLen.Wrapf("expected: %d got: %d", int(pathLen), len(path))
+	}
+	filter.dataPath = string(path)
+	return filter, nil
+}
+
+// ApplyFilter applies the Standard Deviation Filter and returns an
+// outlier list.
+// (i) If more than 1/3 of reveals are corrupted (i.e. invalid json
+// path, invalid bytes, etc.), a corrupt reveals error is returned
+// without an outlier list.
+// (ii) If the number type is invalid, an error is returned without
+// an outlier list.
+// (iii) Otherwise, an outlier list is returned. A reveal is declared
+// an outlier if it deviates from the median by more than the given
+// max sigma. If less than 2/3 of the reveals are non-outliers, "no
+// consensus" error is returned as well.
+func (f FilterStdDev) ApplyFilter(reveals []RevealBody) ([]int, error) {
+	dataList, _, err := parseReveals(reveals, f.dataPath)
+	if err != nil {
+		return nil, err
+	}
+
+	outliers, err := f.detectOutliers(dataList)
+	if err != nil {
+		return outliers, err
+	}
+	return outliers, nil
+}
+
+func (f FilterStdDev) detectOutliers(dataList []string) ([]int, error) {
+	switch f.numberType {
+	case 0x00: // Int32
+		return detectOutliersInteger[int32](dataList, f.maxSigma)
+	case 0x01: // Int64
+		return detectOutliersInteger[int64](dataList, f.maxSigma)
+	case 0x02: // Uint32
+		return detectOutliersInteger[uint32](dataList, f.maxSigma)
+	case 0x03: // Uint64
+		return detectOutliersInteger[uint64](dataList, f.maxSigma)
+	default:
+		return nil, ErrInvalidNumberType
+	}
+}
+
+func detectOutliersInteger[T constraints.Integer](dataList []string, maxSigma uint64) ([]int, error) {
+	var nums []T
+	var corruptCount int
+	var corruptInds []int
+	var z T
+	rt := reflect.TypeOf(z)
+	length := len(dataList)
+	for i, data := range dataList {
+		if data == "" {
+			corruptCount++
+			corruptInds = append(corruptInds, i)
+			continue
+		}
+
+		bz, err := base64.StdEncoding.DecodeString(data)
+		if err != nil {
+			corruptCount++
+			corruptInds = append(corruptInds, i)
+			continue
+		}
+		if len(bz) != int(rt.Size()) {
+			corruptCount++
+			corruptInds = append(corruptInds, i)
+			continue
+		}
+
+		switch rt.Kind() {
+		case reflect.Int32:
+			var num int32
```

Comment:
Really good point. I took this approach because originally I thought we had to use different functions depending on the type, but it seems we can generalize the code and use T.